snapcraft/projects.py

```diff
@@ -0,0 +1,374 @@
+    - version-script (deprecated)
+    """
+
+    name: constr(max_length=40)  # type: ignore
+    title: Optional[constr(max_length=40)]  # type: ignore
+    base: Optional[str]
+    build_base: Optional[str]
+    compression: Literal["lzo", "xz"] = "xz"
+    version: Optional[constr(max_length=32, strict=True)]  # type: ignore
+    contact: Optional[Union[str, UniqueStrList]]
+    donation: Optional[Union[str, UniqueStrList]]
+    issues: Optional[Union[str, UniqueStrList]]
+    source_code: Optional[str]  # XXX: should we validate as URL?
+    website: Optional[str]  # XXX: should we validate as URL?
+    summary: constr(max_length=78)  # type: ignore
+    description: str
+    type: Literal["app", "base", "gadget", "kernel", "snapd"] = "app"
+    icon: Optional[str]
+    confinement: Literal["classic", "devmode", "strict"]
+    layout: Optional[Dict[str, Dict[str, Any]]]
+    license: Optional[str]
+    grade: Literal["stable", "devel"]
+    adopt_info: Optional[str]
+    architectures: List[Architecture] = []
+    assumes: UniqueStrList = []
+    hooks: Optional[Dict[str, Hook]]
+    passthrough: Optional[Dict[str, Any]]
+    apps: Optional[Dict[str, App]]
+    plugs: Optional[Dict[str, Dict[str, str]]]  # TODO: add plug name validation
+    slots: Optional[Dict[str, Dict[str, str]]]  # TODO: add slot name validation
+    parts: Dict[str, Any]  # parts are handled by craft-parts
+    epoch: Optional[int]
+
+    @pydantic.root_validator(pre=True)
+    @classmethod
+    def _validate_mandatory_version(cls, values):
+        if "version" not in values and "adopt-info" not in values:
+            raise ValueError("Snap version is required if not using adopt-info")
+        return values
+
+    @pydantic.root_validator(pre=True)
+    @classmethod
+    def _validate_mandatory_base(cls, values):
+        snap_type = values.get("type")
+        if ("base" in values) ^ (snap_type not in ["base", "kernel", "snapd"]):
+            raise ValueError(
+                "Snap base must be declared when type is not base, kernel or snapd"
+            )
+        return values
+
+    @pydantic.validator("name")
+    @classmethod
+    def _validate_name(cls, name):
+        if not re.match(r"^[a-z0-9-]*[a-z][a-z0-9-]*$", name):
+            raise ValueError(
+                "Snap names can only use ASCII lowercase letters, numbers, and hyphens, "
+                "and must have at least one letter"
+            )
+
+        if name.startswith("-"):
+            raise ValueError("Snap names cannot start with a hyphen")
+
+        if name.endswith("-"):
+            raise ValueError("Snap names cannot end with a hyphen")
+
+        if "--" in name:
+            raise ValueError("Snap names cannot have two hyphens in a row")
+
+        return name
+
+    @pydantic.validator("version")
+    @classmethod
+    def _validate_version(cls, version):
+        if not re.match(r"^[a-zA-Z0-9](?:[a-zA-Z0-9:.+~-]*[a-zA-Z0-9+~])?$", version):
+            raise ValueError(
```

Comment:
what catches these ValueErrors?

Reply:
The pydantic validator (during unmarshal). The errors are converted by pydantic to `pydantic.ValidationError`, which we catch and convert to `ProjectValidationError`. The latter is a `SnapcraftError`.